### DISCUSSION TITLE
Need help with depth peeling interfering with screen shader effect

### DISCUSSION THREAD

**Discussion by bluesillybeard** [bnpr/Malt#528] [Q&A] [answered]
I am working on a fairly advanced outline renderer (Because the builtin one doesn't fit my needs).

When the lines are opaque, all is good (and transparent objects even work correctly when transparency is enabled):
![image](https://github.com/bnpr/Malt/assets/69832658/dea3112e-173c-47ef-808d-4e96ac47fd61)

However, when the outline is transparent, the outline of objects behind is also drawn for some reason:
![image](https://github.com/bnpr/Malt/assets/69832658/8159bedb-0e7a-4f55-a591-a70083499244)

I believe the issue is related to depth peeling, since disabling transparency on all the objects fixes the issue.
Things I have tried so far:
- enable the "transparency single layer" option on all of my objects
- disable the 'Layer Only' option on all of my screen passes
- disable transparency (which worked but I want to have transparent objects)

Here is my project:
[AdvancedMalt2DRenderer_R4.zip](https://github.com/bnpr/Malt/files/13171739/AdvancedMalt2DRenderer_R4.zip)

**Comment by bluesillybeard** [comment]
I found the cause of my problem: the depth peeling (correctly and as expected) blends the outline color buffer:
![image](https://github.com/bnpr/Malt/assets/69832658/f76c85e5-9007-4eb3-a15e-19dbf2c119e8)

**Comment by bluesillybeard** [comment]
I can't figure out a solution that works with the depth peeling implementation, however I can think of a way to make it work by sorting everything from back to front and drawing it with regular Z buffering, so I'm going to close this remarkably unhelpful discussion and start working on a custom render pipeline to do that.

**Comment by pragma37** [comment, marked as answer]
While transparency being blended on top of anything behind is indeed the intended behavior, you can do some workarounds to avoid it:
![imagen](https://github.com/bnpr/Malt/assets/36610999/d524e434-371d-4a5a-aa06-9fa8540d1bdb)

Here what I'm doing is adding 2 extra outputs to the RenderLayers graph. I simply set the line alpha to 1 in a screen shader and store the actual alpha in a separate output. This way lines are composed as opaque, but I can apply the alpha later in the Render graph.
[line-no-alpha-compositing.zip](https://github.com/bnpr/Malt/files/13198852/line-no-alpha-compositing.zip)

It's not the most efficient way to do it, but it works with just nodes. If you want to code a more bespoke solution keep in mind you don't have to go with a fully custom pipeline, you can simply implement your own `RenderLayers` node in a plugin.

BTW, I've updated the line compositing in the Development branch, since I noticed transparent lines were not working correctly.

**Comment by bluesillybeard** [comment]
Making my own ```RenderLayers``` node sounds a lot better, thanks!
